src/java/picard/sam/revertsam/PerReadGroupSorter.java

```diff
@@ -0,0 +1,52 @@
+package picard.sam.revertsam;
+
+import htsjdk.samtools.SAMReadGroupRecord;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.util.PeekableIterator;
+import htsjdk.samtools.util.SortingCollection;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Wraps a collection of sorters, one per read group.
+ * Calling add results in the record being added to the sorter for that read group.   
+ */
+public class PerReadGroupSorter implements RevertSamSorter {
+
+    private final Map<String, SortingCollection<SAMRecord>> sorterMap;
+    private final List<PeekableIterator<SAMRecord>> iterators;
+    
+    PerReadGroupSorter(final Map<String, SortingCollection<SAMRecord>> sorterMap) {
+        this.sorterMap = sorterMap;
+        this.iterators = createIterators(sorterMap.values());
+    }
+
+    private static List<PeekableIterator<SAMRecord>> createIterators(final Collection<SortingCollection<SAMRecord>> sorters) {
+        final List<PeekableIterator<SAMRecord>> iterators = new ArrayList<PeekableIterator<SAMRecord>>();
+        for (final SortingCollection<SAMRecord> sorter : sorters) {
+            final PeekableIterator<SAMRecord> iterator = new PeekableIterator<SAMRecord>(sorter.iterator());
+            iterators.add(iterator);
+        }
+        return iterators;
+    }
+    
+    @Override
+    public void add(final SAMRecord rec) {
+        final SortingCollection<SAMRecord> sorter = getSorter(rec.getReadGroup());
+        sorter.add(rec);
+    }
+
+    private SortingCollection<SAMRecord> getSorter(final SAMReadGroupRecord readGroup) {
```

Comment:
Now that I'm looking at code and not just commenting on comments I agree with you here.